quil-py/src/program/mod.rs

```diff
@@ -0,0 +1,197 @@
+// them directly to an error type here.
+create_exception!(quil, ProgramError, PyValueError);
+create_exception!(quil, ParseError, PyValueError);
+
+py_wrap_type! {
+    #[derive(Debug, PartialEq)]
+    PyProgram(Program) as "Program"
+}
+impl_as_mut_for_wrapper!(PyProgram);
+impl_repr!(PyProgram);
+
+impl Default for PyProgram {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[pymethods]
+impl PyProgram {
+    #[new]
+    pub fn new() -> Self {
+        Self(Program::default())
+    }
+
+    #[staticmethod]
+    pub fn from_string(program: &str) -> PyResult<Self> {
+        Ok(Self(
+            Program::from_str(program).map_err(|e| ParseError::new_err(e.to_string()))?,
+        ))
+    }
+
+    #[getter]
+    pub fn instructions<'a>(&self, py: Python<'a>) -> PyResult<&'a PyList> {
+        Ok(PyList::new(
+            py,
+            self.as_inner()
+                .instructions
+                .iter()
+                .map(|i| i.to_python(py))
+                .collect::<PyResult<Vec<PyInstruction>>>()?,
+        ))
+    }
+
+    #[getter]
+    pub fn calibrations(&self, py: Python<'_>) -> PyResult<PyCalibrationSet> {
+        self.as_inner().calibrations.to_python(py)
+    }
+
+    #[getter]
+    pub fn waveforms(&self, py: Python<'_>) -> PyResult<BTreeMap<String, PyWaveform>> {
+        self.as_inner().waveforms.to_python(py)
+    }
+
+    #[getter]
+    pub fn frames(&self, py: Python<'_>) -> PyResult<PyFrameSet> {
+        self.as_inner().frames.to_python(py)
+    }
+
+    #[getter]
+    pub fn memory_regions(&self, py: Python<'_>) -> PyResult<BTreeMap<String, PyMemoryRegion>> {
+        self.as_inner()
+            .memory_regions
+            .iter()
+            .map(|(name, memory_region)| Ok((name.to_python(py)?, memory_region.to_python(py)?)))
+            .collect()
+    }
+
+    #[getter]
+    // TODO: Should this filtering move to Program? Should we assume memory_regions will always make up all
+    // declarations and simplify this?
+    pub fn declarations(&self, py: Python<'_>) -> PyResult<HashMap<String, PyDeclaration>> {
+        self.as_inner()
+            .to_instructions(true)
+            .iter()
+            .filter_map(|inst| match inst {
+                Instruction::Declaration(declaration) => Some(declaration),
+                _ => None,
+            })
+            .map(|declaration| Ok((declaration.name.clone(), declaration.to_python(py)?)))
+            .collect()
+    }
+
+    #[getter]
+    pub fn defined_gates(&self, py: Python<'_>) -> PyResult<Vec<PyGateDefinition>> {
+        self.as_inner()
+            .to_instructions(true)
+            .iter()
+            .filter_map(|inst| match inst {
+                Instruction::GateDefinition(gate_def) => Some(gate_def.to_python(py)),
+                _ => None,
+            })
+            .collect()
+    }
+
+    fn dagger(&self) -> PyResult<Self> {
```

Comment:
Is there a reason this isn't a public getter?

```suggestion
    #[getter]
    pub fn dagger(&self) -> PyResult<Self> {
```

Reply:
Good catch, it should be `pub` for consistency. I won't add getter because it's not really a property of the `Program`, it returns a transposed copy of the `Program`